### DISCUSSION TITLE
How to render a url as text without an anchor?

### DISCUSSION THREAD

**Discussion by jmwielandt** [remarkjs/remark#1381] [Q&A]
Hi, my markdown states something like...
```md
# Title

Hi, for help you can visit www.example.com
```

I **DO NOT** want it to be an anchor on the html, but the tag is automatically added on render. It is transformed to:
```html
[...]Hi, for help you can visit <a href="www.example.com">www.example.com</a>[...]
```

How can I fix it?

**Comment by jmwielandt** [comment]
i did it!

fix: add `disallowedElements={["a"]} unwrapDisallowed`.

```jsx
<Markdown remarkPlugins={[remarkGfm]} rehypePlugins={[rehypeRaw]} disallowedElements={["a"]} unwrapDisallowed>
  { content }
</Markdown>
```

**Reply by ChristianMurphy** [reply]
@jmwielandt this will disallow any links from appearing

```
www.example.com
<www.example.com>
[www.example.com](www.example.com)
```

will all be unlinked with this approach, I don't think that is your intent?

**Reply by jmwielandt** [reply]
For my case, yes, that's what i wanted. but in other contexts i think would be amazing to have a way to disable just one autolink (i ended up at the remarkGfm repo reading that they don't intend to allow it (globally, at least).

My markdown is being rendered inside an iframe so if you click one link, it tries to load the webpage inside the iframe instead of opening a new tab or substituting the whole page inside the current tab.

**Reply by ChristianMurphy** [reply]
> i think would be amazing to have a way to disable just one autolink

As a general policy I won't recommend creating yet another markdown flavor by disabling only some constructs.
There's a write up on why here: https://github.com/micromark/micromark?tab=readme-ov-file#extending-markdown

But if you choose to make your own flavor you can disable constructs https://github.com/micromark/micromark?tab=readme-ov-file#case-turn-off-constructs

> My markdown is being rendered inside an iframe so if you click one link, it tries to load the webpage inside the iframe instead of opening a new tab or substituting the whole page inside the current tab.

You may be overcomplicating things, `target="_parent"` and `target="_top"` exist if you want links to be handled outside the iframe.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target

**Reply by jmwielandt** [reply]
I just tried with target="_blank" for a new tab but didn't work as expected:

My code:
```md
<a href="www.example.com" target="_blank">www.example.com</a>
```

How it renders:
![image](https://github.com/user-attachments/assets/8ae5c831-4a7c-4c08-949a-9d2e022e64f1)


The HTML:
![image](https://github.com/user-attachments/assets/021c99d8-7d8b-440c-abbe-4695915e29eb)

If i change my code to this:
```md
<a href="http://www.example.com" target="_blank">www.example.com</a>
```

This is the result:
![image](https://github.com/user-attachments/assets/5a8f218f-1c51-42c9-9f1d-a67d3436e14d)

It looks... kinda buggy?

**Reply by ChristianMurphy** [reply]
@jmwielandt there are a lot of things that could be happening.
Could you share a sandbox with a runnable example of what you are seeing?
This is a good starting point: https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-vite

Try to make it a minimal reproducible example https://stackoverflow.com/help/minimal-reproducible-example
remove as many dependencies and customizations as possible while still creating the issue

**Reply by jmwielandt** [reply]
Here it is: https://stackblitz.com/~/github.com/jmwielandt/sdsdf?file=src/app.tsx
i didn't have to add a lot.

**Reply by ChristianMurphy** [reply]
A few questions, is intentional that you are using HTML inside markdown here?
You can attach attributes to markdown ones as well, and they tend to be more intuitive to authors.
In addition HTML is more risky both from a brand perspective (people and add off brand colors) and security perspective (you need to sanitize to prevent JS injection).

The second is how does `iframe` factor in? Your example doesn't include `iframe`.

---

I think you're looking for something more similar to the following

```ts
import Markdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import { visit } from 'unist-util-visit'
import type { Root } from 'hast'

const markdown = `
www.example.com

I need to put a table too...
|a|b|
|-|-|
|1|2|
`;

function rehypeChangeTargetForIframeLinks() {
  return (tree: Root) => {
    visit(tree, "element", (node) => {
      if (node.tagName === 'a') {
        node.properties.target = "_top"
      }
    })
  }
}

function App() {
  return <>
    <Markdown remarkPlugins={[remarkGfm]} rehypePlugins={[rehypeChangeTargetForIframeLinks]}>
      {markdown}
    </Markdown>
  </>
}

export default App;
```

runnable: https://stackblitz.com/edit/github-drwxgy-nhlxsf?file=src%2Fapp.tsx

Where markdown and inline links are supported, but are handled outside the iframe context thanks to the `target="_top"`

**Reply by jmwielandt** [reply]
> is intentional that you are using HTML inside markdown here?

Yes. In my real code I even have css inside ^^u

> You can attach attributes to markdown ones as well, and they tend to be more intuitive to authors.

How can I set the target="_blank" to a md link? i've been searching about it but couldn't find anything.

> In addition HTML is more risky both from a brand perspective (people and add off brand colors) and security perspective (you need to sanitize to prevent JS injection).

Yeah, but in this case is a static markdown file versioned on the same repository.

> The second is how does iframe factor in? Your example doesn't include iframe.

oh, sorry. i thought you were just asking about the anchor issue.

Now i have a full example with the iframe:

https://stackblitz.com/~/github.com/jmwielandt/sdsdf?file=index.html

Your `rehypeChangeTargetForIframeLinks` function works like a charm! thanks! i only changed the "_top" for "_blank".

BUT: the issue of the anchor duplication still exists while creating it with html 🤔 (use case: imagine wanting to have a div with certain style (margins, background-color, etc) inside the anchor, for example).

**Reply by ChristianMurphy** [reply]
> How can I set the target="_blank" to a md link? i've been searching about it but couldn't find anything.

By using the rehype plugin shared above.
The same one you reference in
> Your rehypeChangeTargetForIframeLinks function works like a charm! thanks! i only changed the "_top" for "_blank".

> Yeah, but in this case is a static markdown file versioned on the same repository

For a fully static file, have you considered using a site generator https://mdxjs.com/docs/getting-started/#site-generators or build tool https://mdxjs.com/docs/getting-started/#build-systems ?
To generate the HTML one time, so the browser doesn't need to:
1. download a markdown parser
2. parse and transform all the content
3. render a dynamic react component

on every page load?

> BUT: the issue of the anchor duplication still exists while creating it with html 

/cc @wooorm this looks like a conflict between GFM autolinks and HTML, what is the intended behavior for this.

**Reply by wooorm** [reply]
This is expected. This is how GFM works. It conflicts. Both work: `<a href="https://example.org">https://example.com</a>` turns into `<a href="https://example.org"><a href="https://example.com">https://example.com</a></a>` at first, and then, because `a`s are not allowed to be in each other, into `<a href="https://example.org"></a><a href="https://example.com">https://example.com</a>`

See what URL shows if you hover over this (or inspect the DOM):

```markdown
<a href="https://example.org">https://example.com</a>
```

<a href="https://example.org">https://example.com</a>

The solution is to change your content: people don’t need to read that `http://` and such:

```markdown
<a href="https://example.org" and-other-attributes>example.com</a>
```

Or even better, use descriptive text:

```markdown
<a href="https://example.org" and-other-attributes>some useful label</a>
```

If you want people to read those protocols, or otherwise include raw URLs, use `code` for that:

```markdown
<a href="https://example.org" and-other-attributes>`https://example.com`</a>
```
